c10/mobile/CPUProfilingAllocator.cpp

```diff
@@ -43,12 +51,33 @@ bool validate_allocation_plan(
     }
     auto start_offset = allocation_offsets[alloc_id];
     auto end_offset = allocation_offsets[alloc_id] + event.size;
+    MemBlock mem_block(start_offset, end_offset);
     if (event.type == EventType::Allocate) {
-      if (!allocations.emplace(start_offset, end_offset).second) {
-        return false;
+      // 1. Form a block whose start offset is end_offset of the candidate block.
+      // 2. Find lower_bound. Thus find iterator to the first block whose start offset
+      //    is not less than, i.e. >=, of this block.
+      //    i.e. first_it.start_offset >= mem_block.end_offset
+      //    all blocks after this iterator dont overlap with mem_block
+      // 3. Iterate over all blocks from the beginning till iterator and check
+      //    for overlaps.
+      // Still O(n)
+      const MemBlock query_block(
+          end_offset, std::numeric_limits<uint64_t>::max());
+      auto it = allocations.lower_bound(query_block);
+      if (it != allocations.end()) {
+        while(it != allocations.begin()) {
```

Comment:
you don't need to iterate actually. You just need to check two blocks - one just before (`--it`) and the one just after (`it`). Because the blocks in the set are note overlapping, it's sufficient

Reply:
Yes thats right. I think I forgot the invariant that this set will always only contain non-overlapping intervals. With that invariant you need to check maximum of two other intervals.

Reply:
Also just a clarifying comment: candidate_block = (some_start, some_end) while the block I was using to perform query was (end_offset, max) so the search will point you the first block with start_offset >= some_end. However give the other comment regarding invariant, this does not matter anymore.